subprojects/docs/src/docs/userguide/feature_variants.adoc

```diff
@@ -0,0 +1,226 @@
+
+Most users will only need to care about the first two configurations, to declare the specific dependencies of this feature:
+
+.Declaring dependencies of a feature
+====
+include::sample[dir="java-feature-variant/producer/groovy",files="build.gradle[tags=variant_dependencies]"]
+include::sample[dir="java-feature-variant/producer/kotlin",files="build.gradle.kts[tags=variant_dependencies]"]
+====
+
+[NOTE]
+====
+By convention, Gradle will map the feature name to a capability which group is the same as the group and version as the main component, but a name composed of the main component name and the kebab-cased feature name.
+
+For example, if the group is `org.gradle.demo`, the name of the component is `provider`, its version is `1.0` and the feature is named `mongodbSupport`, so the feature variant will be `org.gradle.demo:provider-mongodb-support:1.0`.
+
+If you choose the capability name yourself or add more capabilities to a variant, it is recommended to follow the same convention.
+====
+
+[[sec:feature_variant_source_set]]
+=== Feature variant source set
+
+In the previous example, we're declaring a feature variant which uses the _main source set_.
+This is a typical use case in the Java ecosystem, where it's, for whatever reason, not possible to split the sources of a project into different subprojects or different source sets.
+Gradle will therefore declare the configurations as described, but will also setup the compile classpath and runtime classpath of the _main source set_ so that it extends from the feature configuration.
+Said differently, this allows you to declare the dependencies specific to a feature in their own "bucket", but everything is still compiled as a single source set.
+There will also be a single artifact (the component Jar) including support for all features.
+
+However, it is often preferred to have a _separate source set_ for a feature. Gradle will then perform a similar mapping, but will _not_ make the compile and runtime classpath of the main component extend from the dependencies of the registered features.
+It will also, by convention, create a `Jar` task to bundle the classes built from this feature source set, using a classifier corresponding to the kebab-case name of the feature:
+
+.Declaring a feature variant using a separate source set
+====
+include::sample[dir="java-feature-variant/producer-separate-sourceset/groovy",files="build.gradle[tags=register_variant]"]
+include::sample[dir="java-feature-variant/producer-separate-sourceset/kotlin",files="build.gradle.kts[tags=register_variant]"]
+====
+
+[[sec::publishing_feature_variants]]
+== Publishing feature variants
+
+[WARNING]
+====
+Depending on the metadata file format, publishing feature variants may be lossy:
+
+- using {metadata-file-spec}[Gradle metadata], everything is published and consumers will get the full benefit of feature variants
```

Comment:
I would move this to be third in the list and add _upcoming_ since it is not yet available in 1.0 and we do not want to convey the wrong message.

Reply:
done